apps/src/EditorAnnotator.js

```diff
@@ -0,0 +1,961 @@
+   */
+  static getBlocksForLine(lineNumber) {
+    let ret = [];
+
+    if (EditorAnnotator.hasBlocks() && EditorAnnotator.isLineBased()) {
+      ret = EditorAnnotator.annotator()?.getBlocksForLine(lineNumber) || [];
+    }
+
+    return ret;
+  }
+
+  /**
+   * Highlights the given line in the active editor.
+   *
+   * @param {number} lineNumber The line number to highlight in the active editor.
+   * @param {string} highlightClass The class name to add to the line.
+   */
+  static highlightLine(lineNumber, highlightClass = 'ace_step') {
+    const info = EditorAnnotator.annotator()?.highlightLine(
+      lineNumber,
+      highlightClass
+    );
+
+    // If there are blocks... highlight line will highlight the blocks for that
+    // line to properly emphasize them.
+    if (EditorAnnotator.hasBlocks()) {
+      // So, before we commit the line, if this is the first highlight, dim all
+      // blocks. That way, the 'highlighted lines' being undimmed are emphasized.
+      if (EditorAnnotator.highlightedLines.length === 0) {
+        // Dim all blocks
+        EditorAnnotator.dimBlocks();
+      }
+
+      // Then undim this line
+      const blocks = EditorAnnotator.getBlocksForLine(lineNumber);
+      for (const block of blocks) {
+        EditorAnnotator.undimBlock(block);
+      }
+    }
+
+    // Remember the line we highlighted
+    if (info) {
+      EditorAnnotator.highlightedLines.push(info);
+    }
+  }
+
+  /**
+   * Clears any lines highlighted by highlightLine.
+   */
+  static clearHighlightedLines() {
+    // Unhighlight each highlighted line in the list and clear the list
+    EditorAnnotator.highlightedLines.forEach(item => {
+      EditorAnnotator.annotator().unhighlightLine(item);
+    });
+    EditorAnnotator.highlightedLines = [];
+
+    // Undim the blocks
+    if (EditorAnnotator.hasBlocks()) {
+      EditorAnnotator.undimBlocks();
+    }
+  }
+
+  /**
+   * Add a line of feedback to the editor.
+   *
+   * @param {number} lineNumber The line number (1-based index)
+   * @param {string} message The text to display next to the line.
+   * @param {string} logLevel The type of annotation ('ERROR', 'INFO', etc)
+   */
+  static annotateLine(lineNumber, message, logLevel = 'INFO') {
+    EditorAnnotator.annotator().annotateLine(lineNumber, message, logLevel);
+  }
+
+  /**
+   * Removes annotations of a particular type.
+   */
+  static clearAnnotations(logLevel = 'INFO') {
+    EditorAnnotator.annotator().clearAnnotations(logLevel);
+  }
+}
+
+if (IN_UNIT_TEST) {
+  /**
+   * Resets state.
+   */
+  EditorAnnotator.reset = function () {
+    EditorAnnotator.studioApp_ = undefined;
+    EditorAnnotator.annotator_ = undefined;
+    EditorAnnotator.code_ = undefined;
+    EditorAnnotator.lines_ = undefined;
+    EditorAnnotator.strippedCode_ = undefined;
+    EditorAnnotator.strippedLines_ = undefined;
+    EditorAnnotator.highlightedLines = [];
+  };
+}
```

Comment:
Since it is singleton-like and its state is assumed to be set-and-forgot once, the tests needed a way to reset that state. That was so closely tied to implementation, that it seemed reasonable to write it near the implementation and expose the logic as a method to the test file. It could be implemented inside the test file, but I try to avoid leaking this level of detail into unit tests.

It is similar to the stubStudioApp/restoreStudioApp functions but I don't explicitly instantiate the singleton, so it's essentially just the restore half.

There was no way to conditionally define the function within the class definition that I know of. So it hangs out on its own in kind of a messy way.